### PR TITLE
Feature users list

### PR DESCRIPTION
**Pull request by christiangelone**
## Summary

 * Adds endpoint for user listing by page number&size
 * Adds tests for authentication of endpoint
 * Adds test for different listing use cases

## Known Issues
No Issues

## Trello Card

https://trello.com/c/KIb1xagA

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -87,3 +89,44 @@ exports.get = (req, res, next) => {
       next(errors.databaseError(err));
     });
 };
+
+exports.list = (req, res, next) => {
+  let query = {
```
Comment (gonziesc) [inline app/controllers/users.js]: const

File: app/controllers/users.js [hunk 2]
```diff
@@ -87,3 +89,44 @@ exports.get = (req, res, next) => {
...
+    page_number: 1,
+    page_size: 100
+  };
+  if (req.body !== null && Object.keys(req.body).length !== 0) {
```
Comment (gonziesc) [inline app/controllers/users.js]: why this?

File: app/controllers/users.js [hunk 3]
```diff
@@ -87,3 +89,44 @@ exports.get = (req, res, next) => {
...
+    query = req.body;
+  }
+
+  return User.count()
+    .then(count => {
+      const pagesQuantity = Math.ceil(count / query.page_size);
+      const pagesLeft = pagesQuantity - query.page_number;
+      const pageOffset = query.page_size * (query.page_number - 1);
+      User.findAll({
```
Comment (gonziesc) [inline app/controllers/users.js]: return

File: app/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const JwtService = require('./../services/jwt'),
+  User = require('../models').User;
+
+exports.secureFor = roles => {
+  return (req, res, next) => {
+    const authHeader = req.headers[JwtService.AUTH_HEADER];
+
+    if (authHeader) {
+      JwtService.decode(authHeader.replace('Bearer ', ''))
```
Comment (gonziesc) [inline app/middlewares/auth.js]: return

File: app/middlewares/auth.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        .then(json => {
+          if (!(roles.indexOf(json.role) > -1)) {
+            return Promise.reject(new Error('Your role is not allowed'));
```
Comment (gonziesc) [inline app/middlewares/auth.js]: next

File: app/middlewares/auth.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  User = require('../models').User;
+
+exports.secureFor = roles => {
+  return (req, res, next) => {
+    const authHeader = req.headers[JwtService.AUTH_HEADER];
+
+    if (authHeader) {
+      JwtService.decode(authHeader.replace('Bearer ', ''))
+        .then(json => {
+          if (!(roles.indexOf(json.role) > -1)) {
```
Comment (gonziesc) [inline app/middlewares/auth.js]: abstract this, is a little bit confusing

File: app/middlewares/auth.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+          }
+          return User.findById(json.id, {
+            attributes: {
+              exclude: ['password']
+            }
+          });
+        })
+        .then(user => {
+          if (user) {
+            req.client = user;
```
Comment (gonziesc) [inline app/middlewares/auth.js]: req.user

File: test/controllers/user.test.js [hunk 8]
```diff
@@ -316,4 +317,249 @@ describe('UserController', () => {
...
+            res.body.overflow.should.be.a('boolean');
+            res.body.overflow.should.equal(false);
+            done();
+          })
+          .catch(err => {
+            done(new Error(`User not be authenticated: ${err.message}`));
+          });
+      });
+    });
+  });
```
Comment (gonziesc) [inline test/controllers/user.test.js:572]: excellent tests my friend

File: test/controllers/user.test.js [hunk 9]
```diff
@@ -316,4 +317,249 @@ describe('UserController', () => {
...
+          done();
+        })
+        .catch(err => {
+          done(new Error(`User not be authenticated: ${err.message}`));
+        });
+    });
+
+    it('Should retrieve a users page list that has one page ahead', done => {
+      const tenPromisesUserCreation = [];
+      for (let index = 0; index < 10; index++) {
```
Comment (gonziesc) [inline test/controllers/user.test.js]: would you abstract this to another place? you are free to call it how you like. we have a script for data creation, that could be a good place

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/controllers/users.js (hunk 2 above)
Comment (christiangelone) [reply app/controllers/users.js]: for some reason req.body = {} when nothing is sent

**Review by gonziesc — APPROVED** [review]

**Pull request merged**